File: server/app/routers/businesses.py
from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.database import get_db
from app.models import Business

router = APIRouter(prefix="/businesses", tags=["businesses"])


class BusinessCreate(BaseModel):
    # Identity
    name: str
    ein: str | None = None
    business_type: str | None = None
    state: str | None = None
    website: str | None = None
    years_in_business: int | None = None

    # Financials
    monthly_revenue: float | None = None
    num_employees: int | None = None
    outstanding_debt: float | None = None

    # Owner
    owner_name: str | None = None
    owner_credit_score: int | None = None


class BusinessResponse(BaseModel):
    id: int
    name: str
    ein: str | None
    business_type: str | None
    state: str | None
    website: str | None
    years_in_business: int | None
    monthly_revenue: float | None
    num_employees: int | None
    outstanding_debt: float | None
    owner_name: str | None
    owner_credit_score: int | None
    is_real: bool | None
    risk_score: float | None
    risk_tier: str | None

    model_config = {"from_attributes": True}


@router.get("/", response_model=list[BusinessResponse])
def list_businesses(
    name: str | None = Query(None, description="Partial name match"),
    state: str | None = Query(None),
    business_type: str | None = Query(None),
    risk_tier: str | None = Query(None, description="low, medium, or high"),
    is_real: bool | None = Query(None),
    min_monthly_revenue: float | None = Query(None),
    max_monthly_revenue: float | None = Query(None),
    db: Session = Depends(get_db),
):
    q = db.query(Business)

    if name is not None:
        q = q.filter(Business.name.ilike(f"%{name}%"))
    if state is not None:
        q = q.filter(Business.state == state)
    if business_type is not None:
        q = q.filter(Business.business_type == business_type)
    if risk_tier is not None:
        q = q.filter(Business.risk_tier == risk_tier)
    if is_real is not None:
        q = q.filter(Business.is_real == is_real)
    if min_monthly_revenue is not None:
        q = q.filter(Business.monthly_revenue >= min_monthly_revenue)
    if max_monthly_revenue is not None:
        q = q.filter(Business.monthly_revenue <= max_monthly_revenue)

    return q.all()


@router.post("/", response_model=BusinessResponse, status_code=201)
def create_business(business: BusinessCreate, db: Session = Depends(get_db)):
    db_business = Business(**business.model_dump())
    db.add(db_business)
    db.commit()
    db.refresh(db_business)
    return db_business

File: frontend/src/api/businesses.ts
const BASE = '/api';

export interface Business {
  id: number;
  name: string;
  ein: string | null;
  business_type: string | null;
  state: string | null;
  website: string | null;
  date_founded: string | null;
  owner_name: string | null;
  created_at: string;
}

export interface Risk {
  id: number;
  monthly_revenue: number | null;
  outstanding_debt: number | null;
  num_employees: number | null;
  owner_credit_score: number | null;
  is_real: boolean | null;
  risk_score: number | null;
  risk_tier: string | null;
  evaluated_at: string;
}

export interface BusinessDetailData extends Business {
  latest_risk: Risk | null;
}

export interface BusinessCreate {
  name: string;
  ein?: string;
  business_type?: string;
  state?: string;
  website?: string;
  date_founded?: string;
  owner_name?: string;
}

export const api = {
  listBusinesses: (): Promise<Business[]> =>
    fetch(`${BASE}/businesses/`).then((r) => r.json()),

  getBusiness: (id: number): Promise<BusinessDetailData> =>
    fetch(`${BASE}/businesses/${id}`).then((r) => r.json()),

  createBusiness: (data: BusinessCreate): Promise<Business> =>
    fetch(`${BASE}/businesses/`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    }).then((r) => r.json()),

  evaluate: (id: number): Promise<Risk> =>
    fetch(`${BASE}/businesses/${id}/evaluate`, { method: 'POST' }).then((r) =>
      r.json(),
    ),

  getRiskHistory: (id: number): Promise<Risk[]> =>
    fetch(`${BASE}/businesses/${id}/risk-history`).then((r) => r.json()),
};
